kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -598,6 +629,23 @@ protected Optional<LogSegment> getLogSegmentForVersion(
                 );
                 throw new IllegalStateException(msg);
             }
+
+            // Reading sidecars only applies for V2 checkpoints (whether with a classic or UUID
+            // naming scheme).
+            if (newCheckpoint.format.usesSidecars()) {
+                final List<FileStatus> sidecarFiles = listSidecars(tableClient);
+                // Only read the checkpoint for sidecars if any sidecars exist in the table.
+                if (!sidecarFiles.isEmpty()) {
+                    // Safe to call .get() here, as we know the CheckpointInstance was initialized
+                    // with a valid filepath.
+                    final Set<Path> referencedSidecars = new HashSet<>(
+                            newCheckpoint.getReferencedSidecars(tableClient, logPath).get());
+                    newCheckpointFileList.addAll(sidecarFiles.stream()
+                            .filter(f -> referencedSidecars.contains(new Path(f.getPath())))
+                            .collect(Collectors.toList()));
+                }
+            }
```

Comment:
can we delay the loading of the sidecard files until we start reading?

Reply:
I think so - will update.